### DISCUSSION TITLE
Accessibility (a11y) in Zed

### DISCUSSION THREAD

**Discussion by iamnbutler** [zed-industries/zed#6576] [Feature Requests]
Opening up this discussion as a place y'all can leave comments, ideas & thoughts about accessibility in Zed.

A11y (accessibility) in Zed will be a long project. Likely lasting far beyond 1.0. Due to GPUI being written from the ground up we don't have access to the same a11y features that Swift, Web-based apps or [insert other language] does.

Making Zed accessible will be a join effort between things on the Zed side, and building out features in GPUI.

What is helpful to hear from you all:
- Specific places in Zed accessibility can be improved.
- Technologies we might use to speed up the implementation of accessible features.
- Examples of apps and products that do accessibility well.

What isn't helpful to hear:
- "You should have used ___ instead of GPUI." Just don't 🙂
- "X theme is not accessible" – If the theme matches it's appearance in other apps (vscode, Atom) this will not be fixed. If Zed's version does not match those themes, please create an issue and I'll take a look.

**Comment by mwcampbell** [comment]
A good first step would be to integrate [AccessKit](https://github.com/AccessKit/accesskit) into GPUI. Feel free to refer to the AccessKit integration in [egui](https://github.com/emilk/egui) for clarification on how to use AccessKit. Let me know if you need help.

**Comment by CurbaiCode** [comment]
I was just thinking that a high contrast mode would be good, but one that can be applied to any theme. This would include things like having all buttons, areas, and boxes have outlines; maybe even a higher contrast version of the UI typefaces.

**Comment by leopepe** [comment]
As a colorblind what helps me in general is having high contrast colours to differentiate the purpose of certain aspects of the UI (options and required to click). In IDEs what is certaily helpful is when the colour schemes contrast green and red, blue and yellow. For instance if the function name colour is in the spectrum of red, the variables will never be a colour in the spetrum of green. (same rule for blue and yellow).

**Comment by mikolysz** [comment]
Screen reader user here. Considering that Zed is being positioned as not just a simple editor but also a collaboration tool, this should be much more of a priority than it seems to be right now. If a team switches to Zed and requires its members to use it down the line, people who rely on accessibility features will be in serious trouble. This is actively going to make some people's lives worse in a very measurable way.

Considering the amount of effort required here, perhaps building a bridge to make collaboration with Zed users work in other editors (like VS Code for example) could be a stopgap measure for 1.0. VS Code is perfectly accessible already and even has some accessibility support for live collaboration built in, so this would be a good compromise if the team isn't willing to invest the resources required for actual accessibility right now.

**Comment by joeldodson** [comment]
Another screen reader here (NVDA on Windows 11) and a admitted VS Code fanboy.
It's encouraging this thread exists, but the idea that accessibility might be along term project that goes beyond the 1.0 release is a discouraging perspective.  You wouldn't say keyboard navigation is a 1.0 feature, then, implicitly not a consideration in general.  Okay, fighting a strawman here, but, to me at least, that's how that comment comes across.
One of the great features of VS Code is their continued active engagement with the blind developer community.
I'm afraid I don't know enough about accessibility APIs, especially outside of Windows, to be helpful technically. I am looking forward though to trying Zed on Windows and hope to engage at a deeper level then. 
I read about the features of Zed and think, that sounds awesome, for the average sighted developer. I was sighted for the first 20 years of my career and appreciate all the features mentioned.  Now I hear those features and am happy for my sighted colleagues, but if they come at the expense of further isolating my blind colleagues, are they worth it?
I'm not suggesting developer tools be limited so we can use them.  That's not constructive for anyone, and I don't think it's necessary.  At the very least though, thinking of accessibility as critical as privacy and security, for example, would be a great position.  I suspect there must be some features passed over because they couldn't be made secure or lacked privacy.  Have any features been passed over because they can't be made accessible?  Has any product owner ever even considered that?
I appreciate Iamn's note to keep comments constructive, and I'm trying to not rant here.  I am though trying to emphasize how frustrating it is when an entire industry, that's not inherently inaccessible, becomes, at least somewhat, inhospitable to a certain community. 
How fast is fast enough?  Is shaving off some milliseconds worth it if a product will be inaccessible?  I appreciate shaving milliseconds on task repeated thousands of times a day adds up.  But can that be done without leveraging the GPU?  Again, my lack of technical depth here might be glaringly obvious, maybe try to read this as a meta-point.  It sounds like engaging the GPU will make accessibility more difficult.  How much speed up do you get by simply using Rust vs Electron?  Is that enough speedup to make a difference?
Again, thanks for this thread and at least considering accessibility early in your journey.  I hope there will be more, and continuous engagement with blind developers going forward.

**Comment by robertmeta** [comment]
I think hooks oriented accessibility should be considered. This is the approach used by Emacspeak which is IMHO leaps and bounds above other accessible interfaces. The reason I think this is a good approach is: 

1. Allows for intelligent adaptation, because a voice stream is not a visual UI, it is an order of magnitude less data dense. 
2. Improves the Zed hooks system as a whole, improving things for sighted and non-sighted users alike. 
3. The majority of the code is OS-neutral, because reading a line aloud on mac, windows, linux should read the same, the only thing that needs to change is the reading system.
4. Can leverage existing voice servers written for Emacspeak, which already supports mac and linux and windows support is coming soon. This tracks well with Zed's roadmap. 
5. Allows UI and TTS interface to both specialize without being hampered by the other side, as they are extremely different experience. For example, handling notification of an edit on another line by a different user. Emacspeak TTS servers already have the idea of notification streams (goes to one ear if enabled). 
 
So, how does it work? It is extremely simple, here is an example of a possible flow. 

1. Zed launched.
2. Zed config contains a "voice server" setting which is a process that is run when Zed starts up
4. Zed sends commands to STDIN of the process, these are simple commands like "tts_say Hello World" would say Hello World to the user.
5. Zed uses the interface to the voice server combined with a series of hooks to handle all TTS interface elements, which means a rich hooks system is required. 
6. Build out the hooks one by one. First two hooks might be "line up" and "line down" each sending that entire line to the TTS server and it is evolutionarily grown. Hook to send char movements, hook to send line joins, etc, etc. Piece by piece until done. 

This approach has happy side-effects too, like:

1. Contributes to shared servers used by multiple editors
2. As said, makes hooks system in Zed more robust. 
3. Once basics are built out, makes it possible for the blind community to do the rest of the development.

**Reply by mikolysz** [reply]
I don't necessarily disagree, but here are a couple important things to note.

1. This solves one use case (screen readers), while integration with native accessibility APIs (e.g. via Accesskit) solves them all. Voice recognition is one notable use case that also relies on accessibility APIs but wouldn't be solved by a hooks system, and there are probably others.
2. I don't think you could do Braille display support this way, or at least not without a lot of pain, and this is something that's very important and a non-starter to many users. Braille displays are usually under exclusive control of the currently running screen reader, and most screen readers don't expose good APIs to manage them. Braille displays aren't just for output, they also let the user move between lines, jump to a specific character with a physical button located above that character's Braille cell etc. If we treat screen readers as dumb programs that should speak what they're told to, instead of giving them full semantic information about the UI, that has no chance of working.
3. Screen readers have specific shortcuts to read specific pieces of information, like "say current word" or "read continuously from the cursor to end of file". If screen readers aren't given the full contents of the document, those shortcuts aren't going to work. This will pollute Zed's keymap with a lot of accessibility-specific commands that users will have to re-learn instead of just using the shortcuts they already know and are used to.
4. Platform conventions. Every platform and every screen reader calls things slightly differently, has a different way of navigating the screen, speaks according to the user's settings, in the voice of the user's choosing, with its own punctuation scheme etc. Trying to replicate this would require a massive amount of work. You could just give up and do things "the Zed way" everywhere, but then you'd have to expose all of these settings in Zed itself.

You should also keep in mind that you'd need hooks for every single thing the editor lets you do, including navigating non-editor UI controls, the entire settings panel, menus, file switching etc. You'd also have to either build in a "micro screen-reader" that would allow reading the static, non-keyboard-navigable elements of the UI, like the current line number, or expose separate keyboard commands for all of them.

Google Docs tried this approach, because web accessibility standards just weren't good enough to implement accessibility for a complex, WYSIWYG rich text editor at the time, and it caused them so much trouble that they had to do a separate accessibility implementation called "Braille mode" that actually does things properly.

**Reply by robertmeta** [reply]
@mikolysz great points.

> This solves one use case (screen readers)... 

Didn't mean to imply it was an "instead of" more of a possible route that is slightly easier on the main editor widget implementation, which right now seems to be a black hole. Didn't mean to imply a "don't implement proper accessibility APIs". 

> I don't think you could do Braille display support this...

Correct, at least on Mac you couldn't. 

> Screen readers have specific shortcuts to read specific pieces of information...

True, but an advanced editor has many movements and editor tooling that are well beyond the primitives provided by the OS editing capacity. This is one of the major features of this appraoch being able to handle things that do not exist in the OS primitives. Multi-user editing events, complex editor operations (multiline join maybe, multiple cursors, etc), this is one of the use cases I specifically was thinking about.

> Platform conventions. Every platform and every screen reader calls things slightly differently, has a different way of navigating the screen, speaks according to the user's settings, in the voice of the user's choosing, with its own punctuation scheme etc. Trying to replicate this would require a massive amount of work. You could just give up and do things "the Zed way" everywhere, but then you'd have to expose all of these settings in Zed itself.

Well, it is work for sure, I wrote the Swiftmac server that does exactly this for Emacspeak (not the bindings, but voice, punctuation, rate, etc). This allows a set of features for heavy text editing and syntax highlighting that translates to voices. From tone changes to complete voice changes for syntax highlighted elements, even rate changes if you want numbers read slower, etc. But you are absolutely right, this is very much making "the Zed way", which seems almost a foregone conclusion when Zed supports text editing primitives and modes (collaboration) that the OS doesn't so the basic text editor won't.

> You should also keep in mind that you'd need hooks for every single thing the editor lets you do, including navigating non-editor UI controls, the entire settings panel, menus, file switching etc. You'd also have to either build in a "micro screen-reader" that would allow reading the static, non-keyboard-navigable elements of the UI, like the current line number, or expose separate keyboard commands for all of them.

Yep, but many things very useful for developers on Mac and Linux are not easily exposed via the screen readers. Windows is far better in this regard. For example, there is no feature that I am aware of on MacOS to get the current line number. I believe VoiceOver simply does not have this. Because screenreaders provide basic primitives there are rather huge gaps for heavy text editing. So either the user with a screen reader doesn't get the feature or you add in support. Line numbers are fairly basic, more advanced stuff is even less supported... and even line numbers I believe on MacOS is missing. 

> Google Docs tried this approach, because web accessibility standards just weren't good enough to implement accessibility for a complex, WYSIWYG rich text editor at the time, and it caused them so much trouble that they had to do a separate accessibility implementation called "Braille mode" that actually does things properly.

... indeed, I basically gave up on Google Docs because of it, I will have to revisit it. TIL. 

----

Again, a lot of great points and this was never a instead of just something that might have value in addition to and depending on the level of hooks could have a lower implementation cost.

**Comment by sempruijs** [comment]
I myself am a screen reader user. Currently I only found one editor that is accesible and implements vim and that is emacspeak. Would love to see accessible vim for zed.

**Reply by deathdogg** [reply]
What editor is that? I am a blind dev and love learning about new tools that work

**Reply by sempruijs** [reply]
Emacspeak is a screen reader for emacs. Emacs only shows text buffers but these buffers are interactive. People made an entire IDE with emacs. Once you have emacspeak installed it works really well. I've used it for a while but vscode brought me back because lsp integration was faster and more responsive.

Btw I wrote a book called coding with eyes closed. Maybe it can help you further on your journey :) 
https://codingwitheyesclosed.com/

**Comment by jscholes** [comment]
It's coming up on a year since the most recent comment on this discussion, and the Windows release of Zed has zero practical accessibility with screen readers (other than system-provided dialogs like file/folder pickers).  Are there are any plans to implement AccessKit as suggested by @mwcampbell in June 2023?

**Comment by frosk1** [comment]
Since I can see "Better Accessibility  (Voice Control and more)" on the Roadmap. Here is something for you to focus on concerning the voice part:


For developers like me with pain issues that uses voice to code, there is already an established standard and a software for voice coding. It is a remarkable project, very well maintaned representing probably the biggest and nearly all communnities for voice coding. It runs on windows, linux and macos.

[Talon](https://talonvoice.com/)
 
To integrate voice coding into an IDE like Zed, you would need to bring the voice coding extenstion called Cursorless into the Zed extenstion space, which uses Talon for voice recognition and a new specific voice coding text editing methodology. 

(_at the moment the extenstion is written in typescript due its original vs code plugin nature, it is already present in neovim and jetbrains IDE plugin systems via quickjs. for zed it would needed to be written in wasm, see here for the cursorless everywhere developer guidelines:_ [cursorless everywhere](https://github.com/cursorless-dev/cursorless/discussions/2189)

[Cusorless](https://www.cursorless.org/)


So concerning accessibility in Zed via voice coding, I gave you the plan:  Just add Cursorless as a Zed extenstion and you have the voice coding accessibility feautre. for the rest of voice control talon users will use your keymaps to create commands they can use with voice in talon to control zed debugger or file tree etc.

! Please do not spend time on developing speech recognition models on your own for this domain, since talon already has put in years of research and development and has the best performing models for this domain and business logic behind it. It is as I said, market standard and used by the voice coding community, so it has the user base already and they will not switch to yours; also since they use it for other apps like browser or office software. you can find multiple examples of projects failing with this, like Serenade !

You are welcome to come to the Talon community slack and you will find all devs of the talon and cursorless projects, as well as nearly all devs with disabilities leveraing voicd to code for their profession. You can introduce yourselves and will be warmly welcomed :)

[talon slack](https://talonvoice.slack.com/join/shared_invite/zt-3a7bgx006-mslQ9DDXpCKqV2slZyArEA#/shared-invite/email)


If you ask me, you can just spend not that much time to add the cursorless extenssion and talk with users and devs on the slack to maybe refine some keymaps in zed for better more efficinet voice control and they bring it into talons. in theory talon could also run native zed commands. And you can ship Zed as an editor for accessibilty support in a next version, which will be a good sounding news for its popularity ;)

But seriously, please consider what I have written here, before running down a rabbit hole you will not like ending up after a while. save yourselves some time.

**Reply by rinOfTheStars** [reply]
Seconding this approach. Pretty sure the main thing holding a third-party extension back here is being able to modify the UI, since that's what Cursorless needs to be able to draw its decorations. Looking at the specs, it seems like that's not possible yet? Seems like a waste to not let extensions use that massive UI library built for this project in any way, even if it's limited compared to what a full app can do with it.

**Comment by warrenjokinen** [comment]
Maybe not accessibility, per se, but I do use the built in "Edit > Speech > Start Speaking" on my Mac to say my writing back to me sometimes. No additional equipment, or costs for this feature.

It helps me proofread my typing a little bit more. I have to leave Zed to do this currently. (VS Code does not offer this either.)

I cut and and paste back and forth from TextEdit. For example, for this message I heard "past" when it needed to be "paste"...

**Comment by warrenjokinen** [comment]
There are literally over a thousand places in the codebase and docs that use **"See", that could be "Refer to", "Use" or something else** even better.

A big undertaking to find all of those, for sure, but a start would be to use better terms now as we move forward.

(I am guilty of saying "Sanity Check", which I found at least 11 times.)

**Comment by mikolysz** [comment]
As a screen reader user, I would kindly suggest prioritizing changes with more user impact.

Please be aware that “see” is just as much a part of the language for many screen reader users as it is for everybody else. It is common for a blind person to say “have you seen …” when referring to an article or (audio described) movie, even when talking to another blind person.

In my opinion, focusing on improving Zed's actual accessibility support would be far more beneficial to downstream users than changing terminology used in its source code, which most of those users will never see. With how few resources are currently spend on accessibility work at Zed, I think a decision to focus on terminology would be misguided at best.

**Reply by warrenjokinen** [reply]
If anywhere in my previous posts I implied that I control any priorities at Zed, then kindly point out exactly where I did that, and I will clarify those posts.

Zed is open source software, therefore both employees and volunteer contributors can make it better. 

I am in the latter category, not an employee, not compensated in any way.

Zed should hire you. They would benefit from your passion.

**Comment by sarimarton-ofsz** [comment]
Hi

I'm not using a screen reader, I'm just using a keyboard... I gave Zed high hopes and started migrating from VSCode because VSCode really started become a perf hog on my work machine. But not having able to navigate through the UI with the keyboard is an instant dealbreaker, and honestly, starting a UI framework like this I find atrocious. Time to time I use full keyboard access in macOS and also Homerow. Specifically, I wanted to access the tiny repository selector in the git pane (my workspace has submodules) and there's no way to use that without the mouse.

**Comment by SeanTolstoyevski** [comment]
Hello @sarimarton-ofsz ,

As a blind user, I couldn't understand this, but are you saying that Zed does not support keyboard navigation and that such a feature doesn't exist?
For example, if I ask the following, do you have any information about them?

1. Can you navigate the file list using the up and down arrow keys, so that the focused file item changes?
2. Does pressing Tab on any control (e.g., treeview, button, etc.) move focus to the next control? Similarly, Shift+Tab should move focus to the previous control.

Zed uses its own UI components and UI management platform entirely. Normally, such events are managed by the OS. Anyway… if Zed's keyboard navigation doesn't support what I've written (these are just examples, keyboard navigation is a vast world), then these issues must be resolved first, before tackling accessibility.

**Reply by sarimarton-ofsz** [reply]
It has a minimum level of tab navigation, but I'd say it's more like a sporadic possibility. You cannot reach all the UI elements, and can't really navigate between regions or even sub-regions with the OS primitives. You can, however move focus to different parts with command palette shortcuts. E.g. you can activate the file explorer pane with Cmd + Shift + e, and you can move the selection with the up and down arrows and load the files with the spacebar.

**Reply by joeldodson** [reply]
Thanks @sarimarton-ofsz  - 

As a formerly sighted developer, with a strong desire to not move my hands from the keyboard, I've often tried to explain the idea of full keyboard navigation to product owners.

When I would reach out to companies about their lack of accessibility on their web apps, I'd get the general response of we value accessibility and have forwarded your comments to our development team ... blah blah blah. I'd try to explain, as a first step, have some sort of minimum commit requirement that all developers must disable any pointing devices and test their changes using only a keyboard. Forget the screen reader, don't even close your eyes. The point is, as you tab/arrow around, do you know where the focus is? Can you access all elements? Can you tell me with certainty what will happen if I reach in and hit space or enter at any time? Do you have access to all the same functionality using only your keyboard as you would with your mouse/trackpad?

If that fails, the PR fails.

There are still the issues of unlabeled elements, lack of alt attributes on graphics, etc. Those might be considered fairly minor though especially with many AI options to have my screen described to me.

If I cannot access an element though without some poining device, I'm screwed.

I tend to find humor in irony. Developers, who are often the most hard core keyboard users, developing a product for developers that is not fully navigable using only the keyboard isn't the funniest irony.

**Comment by dtinth** [comment]
I am unable to use the keyboard to click the [🔄 Reconnect] button on this page:

<img width="1009" height="515" alt="image" src="https://github.com/user-attachments/assets/379c0c7c-5dc7-4512-8a7b-ac436325ae76" />

Pressing [Tab] key does nothing.

## Failed workaround

Normally I use [Homerow](https://www.homerow.app/) to click on buttons I cannot tab to, but none of Zed's button appear on Homerow’s selection

<img width="981" height="491" alt="image" src="https://github.com/user-attachments/assets/2bf8fa2f-9220-4ed6-99ae-57bac238831a" />

## Workaround

My current workaround is to use a separate app, [Superkey](https://superkey.app/) that performs optical character recognition on the screen to allow me to click on it.

<img width="996" height="494" alt="image" src="https://github.com/user-attachments/assets/f0099d9c-c257-440f-8ba0-e3a2ae2a9a43" />

**Comment by paul-geoghegan** [comment]
Hi there, I would definitely love accesskit support and I personally think that it's the best way forward for sure. The accesskit devs seem pretty approachable from what I saw with them reviewing the accesskit PRs for the Godot game engine. I am quite sad to learn of the lack of accessibility though as I was about to install this on my work machine and give it a go.

**Comment by SeanTolstoyevski** [comment]
Hey everyone,

No one has written here yet but Zed seems to be doing some accessibility experiments on the `main` branch. This is currently disabled and you can enable it (set `ZED_EXPERIMENTAL_A11Y=1`)  and try. Some menus, some interface elements are accessible but there is a long way to go.

If you know Rust, you can contribute to Zed in this area because at least how to do it, which method to choose has been determined. Now PRs will be about adding ax support to components; after all, the architecture has been decided.

Still, keep in mind that there is still a long way to go and many things work inconsistently.

I just wrote here because I saw it. I have no responsibility regarding this part of the project, I wanted to inform.

***

On my behalf, many many thanks to the Zed team for taking the time for this.

After the recent VSCode regressions, I had seriously started thinking about writing an IDE for writing Go and C++ - I even wrote part of it as a POC with wxWidgets + scintilla - it became so annoying that VSCode turned into a real bloatware.
